src/System.IO.Pipelines/ReadCursor.cs

```diff
@@ -89,52 +90,23 @@ internal int GetLength(ReadCursor end)
                 return endIndex - startIndex;
             }
 
-            return GetLengthMultiSegment(start, startIndex, endSegment, endIndex);
-        }
-
-        [MethodImpl(MethodImplOptions.NoInlining)]
-        private static int GetLengthMultiSegment(BufferSegment start,
-            int startIndex,
-            BufferSegment endSegment,
-            int endIndex)
-        {
-            var segment = start;
-            var index = startIndex;
-            var length = 0;
-            checked
-            {
-                while (true)
-                {
-                    if (segment == endSegment)
-                    {
-                        return length + endIndex - index;
-                    }
-                    else if (segment.Next == null)
-                    {
-                        return length;
-                    }
-                    else
-                    {
-                        length += segment.End - index;
-                        segment = segment.Next;
-                        index = segment.Start;
-                    }
-                }
-            }
+            return (endSegment.RunningLength - start.Next.RunningLength)
+                   + (start.End - startIndex)
+                   + (endIndex - endSegment.Start);
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        internal ReadCursor Seek(int bytes, ReadCursor end, bool checkEndReachable = true)
+        internal ReadCursor Seek(long bytes, ReadCursor end, bool checkEndReachable = true)
         {
             if (IsEnd)
             {
                 return this;
             }
 
             ReadCursor cursor;
-            if (Segment == end.Segment && end.Index - Index >= bytes)
+            if (Segment == end.Segment && end.Index >= bytes + Index)
             {
-                cursor = new ReadCursor(Segment, Index + bytes);
+                cursor = new ReadCursor(Segment, Index + (int)bytes);
```

Comment:
Maybe add a comment saying how that if `end.Index - Index >= bytes`, bytes must be safely castable to an int. I think this one is a little bit less obvious than the other uncommented int casts.